src/api/utils/tradeQuote/tradequote.ts

```diff
@@ -0,0 +1,404 @@
+      throw new Error('Remaining units too small, incorrectly attempting max');
+    }
+
+    // toToken
+    const positionForToToken = setOnChainDetails
+      .positions
+      .find((p: any) => p.component.toLowerCase() === toTokenAddress.toLowerCase());
+
+    const newToPositionUnits = (positionForToToken !== undefined)
+      ? positionForToToken.unit.add(quoteToUnits)
+      : quoteToUnits;
+
+    const newToUnitsTooSmall = newToPositionUnits.gt(0) && newToPositionUnits.lt(50);
+
+    if (newToUnitsTooSmall) {
+      throw new Error('Receive units too small');
+    }
+  }
+
+  private calculateFromTokenAmount(
+    setOnChainDetails: any,
+    fromTokenAddress: Address,
+    amount: BigNumber
+  ): BigNumber {
+    const positionForFromToken = setOnChainDetails
+      .positions
+      .find((p: any) => p.component.toLowerCase() === fromTokenAddress.toLowerCase());
+
+    if (positionForFromToken === undefined) {
+      throw new Error('Invalid fromToken input');
+    }
+
+    const totalSupply = setOnChainDetails.totalSupply;
+    const impliedMaxNotional = positionForFromToken.unit.mul(totalSupply).div(SCALE);
+    const isGreaterThanMax = amount.gt(impliedMaxNotional);
+    const isMax = amount.eq(impliedMaxNotional);
+
+    if (isGreaterThanMax) {
+      throw new Error('Amount is greater than quantity of component in Set');
+    } else if (isMax) {
+      return impliedMaxNotional.toString();
+    } else {
+      const amountMulScaleOverTotalSupply = amount.mul(SCALE).div(totalSupply);
+      return amountMulScaleOverTotalSupply.mul(totalSupply).div(SCALE);
+    }
+  }
+
+  private tokenDisplayAmount(amount: BigNumber, address: Address): string {
+    return this.normalizeTokenAmount(amount, address).toString();
+  }
+
+  private tokenResponse(address: Address): TokenResponse {
+    const tokenEntry = this.tokenMap[address];
+    return {
+      symbol: tokenEntry.symbol,
+      name: tokenEntry.name,
+      address,
+      decimals: tokenEntry.decimals,
+    };
+  }
+
+  private chainCurrencyAddress(chainId: number): Address {
+    switch (chainId) {
+      case 1:   return '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'; // WETH
+      case 137: return '0x0d500b1d8e8ef31e21c99d1db9a6444d3adf1270'; // WMATIC
+      default: throw new Error(`chainId: ${chainId} is not supported`);
+    }
+  }
+
+  private normalizeTokenAmount(amount: BigNumber, address: Address): number {
+    const tokenScale = BigNumber.from(10).pow(this.tokenMap[address].decimals);
+    return FixedNumber.from(amount).divUnsafe(FixedNumber.from(tokenScale)).toUnsafeFloat();
+  }
+
+  private tokenPriceUsd(amount: BigNumber, address: Address, coinPrices: CoinGeckoCoinPrices): string {
+    const coinPrice = coinPrices[address][USD_CURRENCY_CODE];
+    const normalizedAmount = this.normalizeTokenAmount(amount, address) * coinPrice;
+    return new Intl.NumberFormat('en-US', {style: 'currency', currency: 'USD'}).format(normalizedAmount);
```

Comment:
It's interesting seeing this kind of code here, I know it's a direct mapping of what's happening in the service right now, but it feels like we'd want a separation of generating the data and formatting the data.

We previously followed a "dumb" client type of approach where the client only needed to display springs and not worry about formatting but its weird seeing that now in a library.

Reply:
Now tracking in #65